File: ingest/parse.mjs
/**
 * parse.mjs — S11 + S13. Runs the parser over the corpus, measures it, and gives
 * every document an explicit disposition.
 *
 *   npm run ingest:parse
 *
 * Parsing is delegated to `ingest/parse.py` because pdfplumber is the only tool
 * here that exposes word geometry, and word geometry is what column detection
 * needs. `retrieval-architecture.md` §2 names pdfplumber for this reason.
 *
 * S13's rule: a document that parses badly gets a **stated disposition**, never a
 * silent pass. A page whose text is wrong is worse than a page with no text —
 * empty text retrieves nothing, while spliced text retrieves confidently.
 */

import { execFile, execFileSync } from 'node:child_process';
import { existsSync, readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { availableParallelism } from 'node:os';
import { promisify } from 'node:util';
import { documents, CORPUS_DIR, isMain } from './reconcile.mjs';

const execFileAsync = promisify(execFile);
const CACHE = 'ingest/.cache/parsed';

/**
 * Parsing runs one process per document, several at a time.
 *
 * pdfplumber is slow on the graphics-heavy Carrier product-data books — 16–25 MB
 * files where most of the page is vector drawing. Sequentially the corpus takes
 * long enough that you stop re-running ingestion, and a corpus you avoid
 * re-ingesting drifts from its sources. Criterion 8 also measures full re-ingest
 * wall-clock, so this is part of the deliverable rather than a convenience.
 *
 * Capped below core count: each worker holds a whole PDF's object graph, and the
 * large files here are memory-hungry enough that oversubscribing trades CPU for
 * swap.
 */
const WORKERS = Math.max(2, Math.min(6, availableParallelism() - 2));

/**
 * Below this mean alpha ratio a document is mostly numbers and symbols — a
 * dimensional table rather than prose. Not a failure, but it retrieves poorly and
 * the artifact should say so rather than let it look like a normal document.
 */
const LOW_ALPHA = 0.55;

/** A document with almost no usable page is excluded rather than half-ingested. */
const MIN_USABLE_FRAC = 0.2;

export function parseDocument(doc) {
  const cachePath = join(CACHE, `${doc.file}.json`);
  if (existsSync(cachePath)) return JSON.parse(readFileSync(cachePath, 'utf8'));

  const raw = execFileSync('python', ['ingest/parse.py', join(CORPUS_DIR, doc.file)], {
    encoding: 'utf8',
    maxBuffer: 512 * 1024 * 1024,
  });
  const parsed = JSON.parse(raw);

  mkdirSync(CACHE, { recursive: true });
  writeFileSync(cachePath, JSON.stringify(parsed));
  return parsed;
}

/**
 * Document types that are *supposed* to be mostly numbers.
 *
 * A pressure-temperature chart is a lookup table; measuring its alpha ratio and
 * calling it low-quality penalises a document for being what it is. The three PT
 * charts score 0.23–0.44 and are exactly the right source for a saturation-
 * temperature question — flagging them would have put a caveat on the one
 * document type that answers a whole fault category correctly.
 *
 * Product-data books are different and keep the caveat: they are *mixed* prose and
 * tables, so a low ratio there genuinely does predict weak prose retrieval.
 */
const TABULAR_BY_DESIGN = ['PT Chart'];

/**
 * S13 — one disposition per document, from the measurement. Every document lands
 * in exactly one of these, and the artifact lists all of them.
 */
export function disposition(quality, doc) {
  const usableFrac = quality.page_count ? quality.usable_pages / quality.page_count : 0;
  const tabular = doc && TABULAR_BY_DESIGN.includes(doc.docType);

  if (quality.error_pages === quality.page_count && quality.page_count > 0) {
    return { state: 'excluded', reason: 'every page failed to parse' };
  }
  if (usableFrac < MIN_USABLE_FRAC) {
    return {
      state: 'excluded',
      reason: `only ${quality.usable_pages}/${quality.page_count} pages carry usable text ` +
        `(<${MIN_USABLE_FRAC * 100}%) — image-only or drawing-only document`,
    };
  }
  if (tabular) {
    return {
      state: 'ingest',
      reason: `reference table by design (alpha ${quality.mean_alpha_ratio}) — numeric content is correct here`,
    };
  }
  if (quality.mean_alpha_ratio < LOW_ALPHA) {
    return {
      state: 'ingest-with-caveat',
      reason: `mean alpha ratio ${quality.mean_alpha_ratio} — mixed prose and dimensional tables, ` +
        `expect weak prose retrieval on the table-heavy pages`,
    };
  }
  if (quality.error_pages) {
    return {
      state: 'ingest-with-caveat',
      reason: `${quality.error_pages} page(s) failed to parse and are skipped`,
    };
  }
  return { state: 'ingest', reason: 'clean text layer' };
}

/** Async twin of parseDocument, so several can be in flight at once. */
async function parseDocumentAsync(doc) {
  const cachePath = join(CACHE, `${doc.file}.json`);
  if (existsSync(cachePath)) return JSON.parse(readFileSync(cachePath, 'utf8'));

  const { stdout } = await execFileAsync('python', ['ingest/parse.py', join(CORPUS_DIR, doc.file)], {
    encoding: 'utf8',
    maxBuffer: 512 * 1024 * 1024,
  });
  const parsed = JSON.parse(stdout);
  mkdirSync(CACHE, { recursive: true });
  writeFileSync(cachePath, JSON.stringify(parsed));
  return parsed;
}

export async function parseAll({ onProgress } = {}) {
  const queue = documents();
  const out = new Array(queue.length);
  let next = 0;

  const worker = async () => {
    while (true) {
      const i = next++;
      if (i >= queue.length) return;
      const doc = queue[i];
      const { pages, quality } = await parseDocumentAsync(doc);
      const d = disposition(quality, doc);
      out[i] = { doc, pages, quality, disposition: d };
      onProgress?.(doc, quality, d);
    }
  };

  await Promise.all(Array.from({ length: WORKERS }, worker));
  return out;
}

if (isMain(import.meta.url)) {
  const started = Date.now();
  const results = await parseAll({
    onProgress: (doc, q, d) => {
      const flag = d.state === 'excluded' ? '✗' : d.state === 'ingest-with-caveat' ? '!' : ' ';
      console.log(
        `  ${flag} ${String(q.usable_pages).padStart(4)}/${String(q.page_count).padEnd(4)} pages  ` +
          `${String(q.two_column_pages).padStart(4)} 2-col  α=${q.mean_alpha_ratio.toFixed(2)}  ` +
          `${doc.inScope ? 'in ' : 'out'}  ${doc.label.slice(0, 46)}`
      );
    },
  });

  const sum = (k) => results.reduce((n, r) => n + r.quality[k], 0);
  const by = (s) => results.filter((r) => r.disposition.state === s);

  console.log(`\nS11 — parse quality across ${results.length} documents  (${((Date.now() - started) / 1000).toFixed(1)}s)\n`);
  console.log(`  pages parsed      : ${sum('page_count')}`);
  console.log(`  usable pages      : ${sum('usable_pages')}`);
  console.log(`  low-text pages    : ${sum('low_text_pages')}   (drawings and dimensional tables — nothing to OCR, per A2)`);
  console.log(`  two-column pages  : ${sum('two_column_pages')}   ← each one a splice a naive -layout read would have produced`);
  console.log(`  pages erroring    : ${sum('error_pages')}`);

  console.log(`\nS13 — dispositions`);
  for (const state of ['ingest', 'ingest-with-caveat', 'excluded']) {
    const group = by(state);
    console.log(`  ${state}: ${group.length}`);
    for (const r of group) {
      if (state !== 'ingest') console.log(`      ${r.doc.label.slice(0, 44)} — ${r.disposition.reason}`);
    }
  }
  console.log();
}

File: ingest/parse.py
"""
parse.py — S11 + A2. Per-page text with column awareness, and a quality signal.

    python ingest/parse.py <pdf-path>            -> JSON on stdout

A2 closed the OCR branch: 24 of 25 documents have clean text layers and the
"low-text" Carrier pages are vector dimensional drawings with nothing to OCR. It
repointed S11's quality signal at the real failure instead:

    `pdftotext -layout` merges the two columns of an IOM page onto one line,
    producing fluent nonsense that embeds confidently and that no flag fixes.

That is the failure this file exists to prevent. A spliced line reads like English,
survives every downstream check, and yields a chunk whose text never appeared in
the manual — a citation to a page that does not say what the chunk claims.

So: detect the gutter, read each column separately, and report per page whether a
naive full-width read would have spliced it.
"""

import json
import sys

import pdfplumber

# A gutter must be this fraction of page width to count as a column separator.
MIN_GUTTER_FRAC = 0.035
# Columns are only plausible when the gutter sits near the middle of the page.
GUTTER_BAND = (0.35, 0.65)
# Below this, a page is a diagram or a cover — real text, just not much of it.
LOW_TEXT_CHARS = 400


def find_gutter(words):
    """
    The widest vertical band containing no word, inside the text block.

    Word-occupancy rather than whitespace: a two-column page has a band no word
    crosses, and a single-column page with indented text does not.

    Measured across the **text extent**, not the page box. Searching the page box
    finds the right margin — on a US Letter IOM that is a 54pt band at centre 0.96,
    wider than any real gutter, so every page looked single-column and every
    two-column page would have been spliced. The bug this file exists to prevent,
    reintroduced by the detector meant to catch it.
    """
    if not words:
        return None
    x0 = min(w["x0"] for w in words)
    x1 = max(w["x1"] for w in words)
    width = x1 - x0
    if width <= 0:
        return None

    # 1pt buckets are finer than any real gutter and cheap at this page size.
    cols = int(width) + 1
    occupied = bytearray(cols)
    for w in words:
        a = max(0, int(w["x0"] - x0))
        b = min(cols - 1, int(w["x1"] - x0))
        for i in range(a, b + 1):
            occupied[i] = 1

    best = None
    run_start = None
    for i in range(cols):
        if not occupied[i]:
            if run_start is None:
                run_start = i
        else:
            if run_start is not None:
                run = (run_start, i - 1)
                if best is None or (run[1] - run[0]) > (best[1] - best[0]):
                    best = run
                run_start = None
    if run_start is not None:
        run = (run_start, cols - 1)
        if best is None or (run[1] - run[0]) > (best[1] - best[0]):
            best = run

    if best is None:
        return None

    gap_w = best[1] - best[0]
    centre = (best[0] + best[1]) / 2 / width
    if gap_w < MIN_GUTTER_FRAC * width:
        return None
    if not (GUTTER_BAND[0] <= centre <= GUTTER_BAND[1]):
        return None
    return x0 + (best[0] + best[1]) / 2


def lines_from(words, tol=2.5):
    """Group words into visual lines, then read each left to right."""
    rows = []
    for w in sorted(words, key=lambda w: (round(w["top"], 1), w["x0"])):
        if rows and abs(rows[-1][0]["top"] - w["top"]) <= tol:
            rows[-1].append(w)
        else:
            rows.append([w])
    return [" ".join(x["text"] for x in sorted(r, key=lambda x: x["x0"])) for r in rows]


def parse_page(page):
    words = page.extract_words(use_text_flow=False, keep_blank_chars=False) or []
    gutter = find_gutter(words)
    two_col = gutter is not None

    if two_col:
        left = [w for w in words if w["x1"] <= gutter]
        right = [w for w in words if w["x0"] >= gutter]
        # A word straddling the gutter belongs to whichever side holds more of it.
        for w in words:
            if w["x1"] > gutter > w["x0"]:
                (left if (gutter - w["x0"]) >= (w["x1"] - gutter) else right).append(w)
        text = "\n".join(lines_from(left) + lines_from(right))
    else:
        text = "\n".join(lines_from(words))

    text = text.strip()
    alpha = sum(c.isalpha() for c in text)
    return {
        "page": page.page_number,
        "text": text,
        "chars": len(text),
        "words": len(words),
        "alpha_ratio": round(alpha / len(text), 3) if text else 0.0,
        "two_column": two_col,
        # True when a naive full-width read WOULD have spliced this page. Not a
        # defect — a defect avoided. Reported so the count is visible rather than
        # assumed to be zero.
        "splice_avoided": two_col,
        "low_text": len(text) < LOW_TEXT_CHARS,
    }


def parse(path):
    pages = []
    with pdfplumber.open(path) as pdf:
        for page in pdf.pages:
            try:
                pages.append(parse_page(page))
            except Exception as e:  # one bad page must not lose the document
                pages.append(
                    {
                        "page": page.page_number,
                        "text": "",
                        "chars": 0,
                        "words": 0,
                        "alpha_ratio": 0.0,
                        "two_column": False,
                        "splice_avoided": False,
                        "low_text": True,
                        "error": f"{type(e).__name__}: {e}",
                    }
                )

    usable = [p for p in pages if not p["low_text"]]
    return {
        "pages": pages,
        "quality": {
            "page_count": len(pages),
            "usable_pages": len(usable),
            "low_text_pages": len(pages) - len(usable),
            "two_column_pages": sum(p["two_column"] for p in pages),
            "error_pages": sum("error" in p for p in pages),
            "total_chars": sum(p["chars"] for p in pages),
            "mean_alpha_ratio": round(
                sum(p["alpha_ratio"] for p in usable) / len(usable), 3
            )
            if usable
            else 0.0,
        },
    }


if __name__ == "__main__":
    json.dump(parse(sys.argv[1]), sys.stdout)
